File: python/openral_py/model/object_ref.py
from abc import ABC, abstractmethod
from typing import Any, Dict, Optional, Type


class ObjectRef:
    uid: str
    role: Optional[str]

    def __init__(self, uid: str, role: Optional[str]) -> None:
        self.uid = uid
        self.role = role

    def to_map(self) -> Dict[str, Any]:
        return {
            "UID": self.uid,
            "role": self.role
        }

    @staticmethod
    def from_map(map_: Dict[str, Any]) -> "ObjectRef":
        uid = map_.get("UID")
        role = map_.get("role")

        if uid is None:
            raise ValueError("UID is required")
        elif not isinstance(uid, str):
            raise TypeError("UID must be a string")

        return ObjectRef(uid, role)

File: flutter/openral_flutter/lib/model/object_ref.dart
class ObjectRef {
  final String uid;
  final String? role;

  ObjectRef({
    required this.uid,
    this.role,
  });

  Map<String, dynamic> toMap() {
    return {
      'UID': uid,
      'role': role,
    };
  }

  static ObjectRef fromMap(Map<String, dynamic> map) {
    dynamic uid = map['UID'];

    if (uid == null) {
      throw ArgumentError('UID is required');
    } else if (uid is! String) {
      throw ArgumentError('UID must be a string');
    }

    return ObjectRef(uid: uid, role: map['role']);
  }
}
